internal/service/ipam/api_next_available_ip_data_source.go

```diff
@@ -98,54 +107,285 @@ func (d *NextAvailableIPDataSource) Schema(_ context.Context, _ datasource.Schem
+	var (
+		apiRes *ipam.NextAvailableIPResponse
+		err    error
+	)
+
+	switch id[:strings.LastIndex(id, "/")] {
 	case "ipam/address_block":
 		apiRes, _, err = d.client.IPAddressManagementAPI.AddressBlockAPI.
-			ListNextAvailableIP(ctx, data.Id.ValueString()).
-			Count(int32(data.Count.ValueInt64())).
-			Contiguous(data.Contiguous.ValueBool()).
+			ListNextAvailableIP(ctx, id).
+			Count(int32(count)).
+			Contiguous(contiguous).
 			Execute()
 
 	case "ipam/subnet":
 		apiRes, _, err = d.client.IPAddressManagementAPI.SubnetAPI.
-			ListNextAvailableIP(ctx, data.Id.ValueString()).
-			Count(int32(data.Count.ValueInt64())).
-			Contiguous(data.Contiguous.ValueBool()).
+			ListNextAvailableIP(ctx, id).
+			Count(int32(count)).
+			Contiguous(contiguous).
 			Execute()
 
 	case "ipam/range":
 		apiRes, _, err = d.client.IPAddressManagementAPI.RangeAPI.
-			ListNextAvailableIP(ctx, data.Id.ValueString()).
-			Count(int32(data.Count.ValueInt64())).
-			Contiguous(data.Contiguous.ValueBool()).
+			ListNextAvailableIP(ctx, id).
+			Count(int32(count)).
+			Contiguous(contiguous).
 			Execute()
 	}
 
+	return apiRes, err
+}
+
+func (d *NextAvailableIPDataSource) findNextAvailableIPsByTags(ctx context.Context, data IpamsvcNextAvailableIPModel) ([]ipam.Address, error) {
+	tagFilterStr := flex.ExpandFrameworkMapFilterString(ctx, data.TagFilters, nil)
+
+	resourceType := data.ResourceType.ValueString()
+	var resources []string
+	var err error
+
+	// Get resources matching tags
+	switch resourceType {
+	case "address_block":
+		resources, err = d.findAddressBlocksByTags(ctx, tagFilterStr)
+	case "subnet":
+		resources, err = d.findSubnetsByTags(ctx, tagFilterStr)
+	case "range":
+		resources, err = d.findRangesByTags(ctx, tagFilterStr)
+	}
+
 	if err != nil {
-		resp.Diagnostics.AddError("Client Error", fmt.Sprintf("Unable to read Address, got error: %s", err))
-		return
+		return nil, err
 	}
-	data.FlattenResults(ctx, apiRes.GetResults(), &resp.Diagnostics)
 
-	// Save updated data into Terraform state
-	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+	if len(resources) == 0 {
+		return nil, fmt.Errorf("no %ss found with the given tags", resourceType)
+	}
+
+	// Try to get next available IPs from each resource
+	var allAddresses []ipam.Address
+	count := int64(data.Count.ValueInt64())
+	contiguous := data.Contiguous.ValueBool()
+
+	for _, resourceID := range resources {
+		// First check if this resource has at least one available IP
+		checkRes, err := d.getNextAvailableIPsByID(ctx, resourceID, 1, contiguous)
+		if err != nil || len(checkRes.GetResults()) == 0 {
+			continue
+		}
+
+		// Try to get as many IPs as needed from this resource
+		remainingCount := int64(count) - int64(len(allAddresses))
+
+		// Start with requesting the full remaining count
+		apiRes, err := d.getNextAvailableIPsByID(ctx, resourceID, remainingCount, contiguous)
```

Comment:
```suggestion
		apiRes, apiErr := d.getNextAvailableIPsByID(ctx, resourceID, remainingCount, contiguous)
```
Error is getting foreshadowed by the one declared on line 235.